File: src/KubeOps.Transpiler/Entities.fs
module KubeOps.Transpiler.Entities

open System
open System.Reflection
open KubeOps.Abstractions.Entities
open KubeOps.Abstractions.Entities.Attributes
open k8s.Models

let private defaulted (value: string) (defaultValue: string) =
    if String.IsNullOrWhiteSpace(value) then
        defaultValue
    else
        value

/// <summary>
/// Converts the given type to an <see cref="EntityMetadata"/> object.
/// </summary>
/// <param name="entityType">The type to convert.</param>
/// <returns>The <see cref="EntityMetadata"/> object representing the type.</returns>
/// <exception cref="ArgumentException">Thrown if the given type is not a valid Kubernetes entity.</exception>
let ToEntityMetadata (entityType: Type) =
    match
        (entityType.GetCustomAttribute<KubernetesEntityAttribute>(),
         entityType.GetCustomAttribute<EntityScopeAttribute>())
    with
    | null, _ -> raise <| ArgumentException "The given type is not a valid Kubernetes entity."
    | attr, scope ->
        (EntityMetadata(
            defaulted attr.Kind entityType.Name,
            defaulted attr.ApiVersion "v1",
            (if String.IsNullOrWhiteSpace(attr.Group) then
                 null
             else
                 attr.Group),
            (if String.IsNullOrWhiteSpace(attr.PluralName) then
                 null
             else
                 attr.PluralName)
         ),
         match scope with
         | null -> EntityScope.Namespaced.ToString()
         | _ -> scope.Scope.ToString())

File: src/KubeOps.Cli/Transpilation/Entities.cs
using System.Reflection;

using k8s.KubeConfigModels;
using k8s.Models;

using KubeOps.Abstractions.Entities;
using KubeOps.Abstractions.Entities.Attributes;

namespace KubeOps.Cli.Transpilation;

internal static class Entities
{
    public static (EntityMetadata Metadata, string Scope) ToEntityMetadata(MetadataLoadContext context, Type entityType)
        => (entityType.GetCustomAttributeData<KubernetesEntityAttribute>(),
                entityType.GetCustomAttributeData<EntityScopeAttribute>()) switch
            {
                (null, _) => throw new ArgumentException("The given type is not a valid Kubernetes entity."),
                ({ } attr, var scope) => (new(
                        Defaulted(
                            attr.GetCustomAttributeNamedArg<string>(nameof(KubernetesEntityAttribute.Kind)),
                            entityType.Name),
                        Defaulted(
                            attr.GetCustomAttributeNamedArg<string>(nameof(KubernetesEntityAttribute.ApiVersion)),
                            "v1"),
                        attr.GetCustomAttributeNamedArg<string>(nameof(KubernetesEntityAttribute.Group)),
                        attr.GetCustomAttributeNamedArg<string>(nameof(KubernetesEntityAttribute.PluralName))),
                    scope switch
                    {
                        null => Enum.GetName(EntityScope.Namespaced) ?? "namespaced",
                        _ => Enum.GetName(
                            scope.GetCustomAttributeCtorArg<EntityScope>(context, 0)) ?? "namespaced",
                    }),
            };

    private static string Defaulted(string? value, string defaultValue) =>
        string.IsNullOrWhiteSpace(value) ? defaultValue : value;
}
